[mobile] Mobile RPCs

halseth opened the pull request:
This PR adds the necessary APIs and build scripts in order to compile `lnd` for mobile platforms.

## Building mobile libraries

### Prerequisites 
#### gomobile
Follow [gomobile](https://github.com/golang/go/wiki/Mobile) in order to intall `gomobile` and dependencies.

Remember to run `gomobile init`! (otherwise the `lnd` build might just hang).

#### falafel
Install [`falafel`](https://github.com/halseth/falafel):
```
go get -u -v github.com/halseth/falafel
``` 

### Building `lnd` for iOS
```
make ios
```

### Building `lnd` for Android
```
make android
```

`make mobile` will build both iOS and Android libs.

### Libraries
After the build has succeeded, the libraries will be found in `mobile/build/ios/Lndmobile.framework` and `mobile/build/android/Lndmobile.aar`. Reference your platforms' SDK documentation for how to add the library to your project.

### API docs
TODO

ottosuess commented on mobile/bindings.go:
```diff
@@ -0,0 +1,39 @@
+func Start(extraArgs string, callback Callback) {
+	// Add the extra arguments to os.Args, as that will be parsed during
+	// startup.
+	os.Args = append(os.Args, strings.Fields(extraArgs)...)
```
splitting the `extraArgs` string breaks if I want to pass a path that contains a space.
e.g. `/Library/Application\ Support`

halseth replied:
Can you check if this helps: https://github.com/lightningnetwork/lnd/pull/3282/commits/22f7b0c44ab0bdc2cca7eedcaa73cb4b0386a3e0

cfromknecht commented:
@halseth didn't get a chance to review the autogenerated code :(

halseth commented:
@cfromknecht added a temp commit with the generated API!

ottosuess commented on mobile/router_api_generated.go:
```diff
@@ -0,0 +1,237 @@
+// NOTE: This method produces a stream of responses, and the receive stream can
+// be called zero or more times. After EOF error is returned, no more responses
+// will be produced.
+func SendPayment(msg []byte, rStream RecvStream) {
```
same name & package as the `SendPayment` method in lightning_api_generated.go.
maybe add a prefix to the methods generated by sub-services?

sergioabril replied:
Agree. But as @halseth suggested somewhere else, you can try setting `usePrefex=1` on `gen_bindings.sh`

(Haven't tried though, I manually changed names)

halseth replied:
Made it possible to add directly to make (https://github.com/lightningnetwork/lnd/pull/3282/commits/73cb325045c3ec874fbf31604254127be0f565b2):
```
make ios tags="routerrpc" prefix=1
```

ottosuess commented:
Would be cool to have a way to inject custom protoc plugins like protoc-gen-swift.

At zap we have a custom "falafel" plugin to generate swift apis that i'd like to inject somehow.

wbobeirne commented on mobile/gen_bindings.sh:
```diff
@@ -0,0 +1,61 @@
+
+# Generate APIs by passing the parsed protos to the falafel plugin.
+opts="package_name=$pkg,target_package=$target_pkg,listeners=lightning=lightningLis walletunlocker=walletUnlockerLis,mem_rpc=1"
+protoc -I/usr/local/include -I. \
```
Just thought I'd mention that I didn't have this command, had to run `brew install protobuf` to get it. Might be worth adding it as a prerequisite to the docs.

halseth commented:
> Would be cool to have a way to inject custom protoc plugins like protoc-gen-swift.
> 
> At zap we have a custom "falafel" plugin to generate swift apis that i'd like to inject somehow.

I think this is a pro-level feature for now, so can be done by modifying `gen_bindings` directly 😄

Roasbeef commented on mobile/bindings.go:
```diff
@@ -0,0 +1,58 @@
+	"github.com/lightningnetwork/lnd"
+)
+
+// Start starts lnd in a new goroutine.
```
How is `lnd` meant to be torn down? Or is the assumption everything will die as soon as the parent process does? 

Is it possible for us to receive signals on the mobile systems as we do right now on the current base operating systems? Certain sub-systems like the tower client rely on being notified for a critical shutdown vs a regular one.

Roasbeef commented on mobile/bindings.go:
```diff
@@ -0,0 +1,58 @@
+	// We call the main method with the custom in-memory listeners called
+	// by the mobile APIs, such that the grpc server will use these.
+	cfg := lnd.ListenerCfg{
+		WalletUnlocker: walletUnlockerLis,
```
Commit ordering wise, these variables don't yet exist.

halseth replied:
Variables are created by the build script from the previous commit.

Roasbeef commented on mobile/bindings.go:
```diff
@@ -0,0 +1,58 @@
+		}
+	}()
+
+	// TODO(halseth): callback when RPC server is running.
```
Care to elaborate on this?

halseth replied:
Added more to the todo: https://github.com/lightningnetwork/lnd/pull/3282/commits/40d433ce26fee479b72e42227fb06fe5dcf1ad69

Roasbeef commented on lnd.go:
```diff
@@ -232,13 +232,49 @@ func Main() error {
 		params, err := waitForWalletPassword(
-			cfg.RPCListeners, cfg.RESTListeners, serverOpts,
-			restDialOpts, restProxyDest, tlsCfg,
+			cfg.RPCListeners, cfg.RESTListeners, restDialOpts,
```
Do we still need to pass in the raw listeners if we have this new function closure? Also will REST on mobile be disabled by default?

halseth replied:
Removed passed in listeners and made function type: https://github.com/lightningnetwork/lnd/pull/3282/commits/04870f136bec3bf40e85bd5c3956a876db730364

Rest must be enabled by setting `restlisten` in the config on mobile as well.

Roasbeef commented on rpcserver.go:
```diff
@@ -441,10 +447,11 @@ var _ lnrpc.LightningServer = (*rpcServer)(nil)
+	subServerCgs *subRPCServerConfigs, restDialOpts []grpc.DialOption,
+	restProxyDest string, atpl *autopilot.Manager,
+	invoiceRegistry *invoices.InvoiceRegistry, tower *watchtower.Standalone,
+	tlsCfg *tls.Config, getListeners func() ([]net.Listener, func(),
```
This is a bit cluttered now, perhaps we should make a new type for this function closure type?

halseth replied:
Made function closure type: https://github.com/lightningnetwork/lnd/pull/3282/commits/04870f136bec3bf40e85bd5c3956a876db730364

Roasbeef commented on lnd.go:
```diff
@@ -93,10 +93,22 @@ var (
+
+	// RPCListener can be set to the listener to use for the RPC server. If
+	// nil a regular network listener will be created.
+	RPCListener net.Listener
```
Do we have any example code w.r.t how these are set and initialized in a mobile setting?

halseth replied:
These are generated by `falafel` on a per-service basis, as seen here: https://github.com/lightningnetwork/lnd/pull/3282/commits/0fb59d1d845276dbb6fafa55b4d3aeeab3bae856#diff-d258038f472562e6320fe2318a48ec8fR19

halseth commented on mobile/autopilot_api_generated.go:
```diff
@@ -0,0 +1,163 @@
+			req proto.Message) (proto.Message, error) {
+
+			// Get the gRPC client.
+			client, close, err := getAutopilotClient()
```
close is a reserved keyword

halseth commented:
This commit currently breaks the build script: https://github.com/golang/mobile/commit/597adff16ade9d88626f8caea514bb189b8c74ee

So need to stay on an older revision for now to successfully build.

halseth commented on lnd.go:
```diff
@@ -232,13 +244,60 @@ func Main() error {
+		return grpcListeners, cleanup, serverOpts, nil
+	}
+
+	// walletUnlockerListeneres is a closure we'll hand to the wallet
```
Fixed: https://github.com/lightningnetwork/lnd/pull/3282/commits/85c2063ec57c23ffc3b6182f8632528895cd5a8e

halseth commented on mobile/bindings.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+// Start starts lnd in a new goroutine.
+//
```
Assumption is that it will die with the parent process (it is actually the same process IIRC). It can also be stopped by calling the `Stop` RPC.

Haven't investigated if we can rely on the signals on mobile platforms. Can be investigated along with tower support.

halseth commented on mobile/bindings.go:
```diff
@@ -0,0 +1,58 @@
+// extraArgs can be used to pass command line arguments to lnd that will
+// override what is found in the config file. Example:
+//	extraArgs = "--bitcoin.testnet --lnddir=\"/tmp/folder name/\" --profile=5050"
+func Start(extraArgs string, callback Callback) {
```
The type `[]string`  is not supported across the gomobile boundary: https://godoc.org/golang.org/x/mobile/cmd/gobind#hdr-Type_restrictions

halseth commented on lnd.go:
```diff
@@ -232,13 +244,60 @@ func Main() error {
+		[]grpc.ServerOption, error) {
+
+		// If we have chosen to start with a dedicated listener for the
+		// wallet unlocker, we we return it directly, and empty server
```
Fixed: https://github.com/lightningnetwork/lnd/pull/3282/commits/f37d8a9f8c5f998679e4922d11ccd04355853cfa

halseth commented:
> This commit currently breaks the build script: [golang/mobile@597adff](https://github.com/golang/mobile/commit/597adff16ade9d88626f8caea514bb189b8c74ee)
> 
> So need to stay on an older revision for now to successfully build.

Latest `gomobile@HEAD` fixes this issue.

halseth commented on mobile/sample_lnd.conf:
```diff
@@ -0,0 +1,13 @@
+[Application Options]
+debuglevel=info
+no-macaroons=1
```
The created bindings doesn't support macaroons. We could add it, but I felt it would be unnecessary, since it is all in-process communication anyway.

halseth commented:
PTAL @Roasbeef @valentinewallace

halseth commented on lnd.go:
```diff
@@ -93,6 +93,10 @@ var (
 )
 
+// rpcListeners us a function type used for closures that fetches a set of RPC
+// listeners for the current configuration.
```
Not totally sure what you mean, but expanded the comment to note what the default return values should be: https://github.com/lightningnetwork/lnd/pull/3282/commits/78bc0b1128edc56f5f7d095a66e3a025d6a2d89e

halseth commented:
Rebased and removed generated code. Generated code can still be inspected here: https://github.com/lightningnetwork/lnd/tree/78bc0b1128edc56f5f7d095a66e3a025d6a2d89e/mobile

valentinewallace commented on mobile/gen_bindings.sh:
```diff
@@ -0,0 +1,65 @@
+    use_prefix="1"
+fi
+
+
```
Nit: extra newline possibly?

halseth replied:
Fixed: https://github.com/lightningnetwork/lnd/pull/3282/commits/1a11c0a95227b5b5e46fae3eeba30761dfe09f36

valentinewallace commented on lnd.go:
```diff
@@ -93,10 +93,29 @@ var (
+	RPCListener net.Listener
+}
+
+// rpcListeners us a function type used for closures that fetches a set of RPC
```
nit: `s/rpcListeners us/rpcListeners is`

halseth replied:
Fixed: https://github.com/lightningnetwork/lnd/pull/3282/commits/a641da4d9571fde5684f3b378e3ee3963bfc16e8

halseth commented on mobile/README.md:
```diff
@@ -0,0 +1,55 @@
+#### Generating proto definitions for your language.
+In order to call the methods in the generated library, the serialized proto for the given RPC call must be provided. Similarly, the response will be a serialized proto.
+
+I order to generate protobuf definitions for your language of choice, add the proto plugin to the `protoc` invocations found in [`gen_protos.sh`](../lnrpc/gen_protos.sh). For instance to generate protos for Swift, add `--swift_out=.` and run `make rpc`.
```
Fixed: https://github.com/lightningnetwork/lnd/pull/3282/commits/210eeb82f949c99781c08f286db5d037a51dcc3b

halseth commented:
Should now be ready to merge
